### PR TITLE
Bootstrap the widgets screen design

### PR DESCRIPTION
**Pull request by youknowriad**
Related #13204 

Capture d’écran 2019-03-25 à 9 42 19 AM

This PR bootstraps the widgets screen design:

 - Saving/fetching... doesn't work
 - It uses a static list of areas
 - It cherry-picks #14369 in order to have separate states for each widget area
 - It uses the block editor module
 - It mimics the aria region navigation of the edit post screen

This PR doesn't intent to solve all the issues related to this screen and is intended to land a first dumb screen in order to allow us to parallelize work and start tackling all the related issues.

**Testing instructions**

 - Navigate to `/wp-admin/admin.php?page=gutenberg-widgets`
 - You should see what ressembles the screenshot above.

### CODE REVIEW

File: packages/edit-widgets/src/components/header/style.scss
Comment (youknowriad) [inline packages/edit-widgets/src/components/header/style.scss]: A lot of these styles to properly style the sidebar and the top bar and the main area are shared with the `edit-post` module. While I don't mind duplication as it's very related to the screen (those screens are similar) but I think we could extract some reusable mixins here (maybe a follow-up).

**Comment by jasmussen** [comment]
Nice. Just looking at the screenshot, this might be a good opportunity to get work going on the generic appender. It really feels like there should be a "plus" button instead of the text-based appender. Here's a recent mockup I made for the column PR (#14241, CC @getdave) that explores the same:

Screenshot 2019-03-25 at 09 53 35

**Comment by mapk** [comment]
> this might be a good opportunity to get work going on the generic appender. It really feels like there should be a "plus" button instead of the text-based appender.

I agree. This was also proposed in the original issue. I think Riad is working his way there most likely. 😉 

https://github.com/WordPress/gutenberg/issues/13204#issuecomment-459175233

**Comment by youknowriad** [comment]
yes, actually, there's a lot more to be done here in addition to the appender change. The idea of this PR is to provide the basis in order to allow us to fix all these issues in multiple parallel PRs.

Personally, the only blocker I see in this PR is #14369. It's cherry-picked here but should be solved first and then we can merge this and iterate.

**Review by aduth — COMMENTED** [review]
Code looks good, particularly mindful of scaffolding nature of the pull request 👍 I'll give it a more thorough shake tomorrow, assuming via https://github.com/WordPress/gutenberg/pull/14369#discussion_r268179596 there's more work to be done with the parent/child store implementation.

File: packages/edit-widgets/src/components/widget-area/index.js
Comment (youknowriad) [reply packages/edit-widgets/src/components/widget-area/index.js]: This requires #14369 so I was just waiting for it.

File: packages/edit-widgets/src/components/sidebar/style.scss
Comment (youknowriad) [reply packages/edit-widgets/src/components/sidebar/style.scss]: I don't really know, this is copied from the edit-post but I can test.

Comment (youknowriad) [reply packages/edit-widgets/src/components/sidebar/style.scss]: This looks like an explicit change done for mobile here https://github.com/WordPress/gutenberg/pull/1691

In my testing it doesn't impact to remove but given how fragile these things are, I prefer to leave and defer to @jasmussen

File: packages/edit-widgets/src/components/widget-area/index.js
Comment (youknowriad) [reply packages/edit-widgets/src/components/widget-area/index.js]: I'm going to open a PR for this separately.

Comment (youknowriad) [reply packages/edit-widgets/src/components/widget-area/index.js]: Done in https://github.com/WordPress/gutenberg/pull/14678

File: packages/data/src/registry.js
Comment (noisysocks) [inline packages/data/src/registry.js]: Why subscribe to the parent twice? Looks like this might have come from a bad merge.

File: packages/edit-widgets/src/components/widget-area/index.js
Comment (noisysocks) [reply packages/edit-widgets/src/components/widget-area/index.js]: Should we just remove the rule? It doesn't [bring me joy](https://twitter.com/dan_abramov/status/1086215771401920512) 😛

File: packages/edit-widgets/src/components/layout/index.js
Comment (noisysocks) [inline packages/edit-widgets/src/components/layout/index.js]: Time for our favourite past time... ✨ _naming things!_ ✨

It looks like WordPress calls _the part of the page where a group of widgets appears_ a ["sidebar" in the code](https://developer.wordpress.org/reference/functions/register_sidebar/) but a ["Widget Area" in the documentation](https://codex.wordpress.org/Widgets_API#Displaying_Widgets_and_Widget_Areas). Which term should we use in Gutenberg?

I am thinking it would be confusing, after Phase 2 is merged into `trunk`, to have both `wp.data.select( 'core/edit-widgets' ).getWidgetAreas()` and `register_sidebar()`.

(Not a blocker for this PR—just trying to get ahead of the terminology here.)

**Review by noisysocks — COMMENTED** [review]
Tested and it looks like a good stub 👍 — will be useful for https://github.com/WordPress/gutenberg/pull/14251!

File: packages/data/src/registry.js
Comment (youknowriad) [reply packages/data/src/registry.js]: Good catch :)

File: packages/edit-widgets/src/components/layout/index.js
Comment (youknowriad) [reply packages/edit-widgets/src/components/layout/index.js]: Yeah, I don't really know. I don't like "sidebars" personally as it's not really what these things are meant for but I can be on board with it for consistency. @aduth our naming expert to the rescue? :P

**Comment by youknowriad** [comment]
I think this is not ready to land to be able to work in parallel after.

File: packages/edit-widgets/src/components/header/style.scss
Comment (aduth) [inline packages/edit-widgets/src/components/header/style.scss]: Possible future task: Is there sense in making this specific to "editor"? Or is it more about how the UI interacts with the admin sidebar more generally?

File: packages/edit-widgets/src/components/layout/index.js
Comment (aduth) [reply packages/edit-widgets/src/components/layout/index.js]: The help text in the current widgets screen also includes some references to this concept as a "widget area", so there's certainly precedent for both cases. I assume it's likely a matter of historical legacy on the "sidebar" naming.

image

Phase 2 could be a good opportunity to push to adopt the "area" naming, and move away from considering these as constrained to sidebars. Consistency with existing function naming is worth considering, but I also wonder if it's not out-of-the-question to consider a soft-deprecation of existing `register_sidebar` functions in favor of newly-introduced equivalent `register_widget_area`, etc. functions? It would need to be a conscious and agreed-upon effort, but I could definitely get behind "area" naming if that were an option.

In any case, I also agree that it needn't block merging this, especially as they're not exposed as part of a public interface (aside from the one class name).

File: lib/widgets-page.php
Comment (aduth) [inline lib/widgets-page.php]: This should include a terminating semi-colon. Otherwise it may result in other `wp_add_inline_script` for `wp-edit-widgets` being treated as a continuation of this code, e.g. an IIFE would result in "VM662:2 Uncaught TypeError: wp.editWidgets.initialize(...) is not a function".

**Review by aduth — APPROVED** [review]

**Comment by aduth** [comment]
I observe local changes to `package-lock.json` as a result of the `edit-widgets` dependency changes. Those should be committed.

**Comment by aduth** [comment]
>I think this is not ready to land to be able to work in parallel after.

I assume you meant (and had previously read this as) "now ready"... which has a very different meaning from a literal interpretation of the text 😄

File: packages/edit-widgets/src/components/layout/index.js
Comment (mapk) [reply packages/edit-widgets/src/components/layout/index.js]: There was some conversation about this in the [customizer issue](https://github.com/WordPress/gutenberg/issues/13205) as well. I went with "Block Areas" in the [prototype](https://wp.invisionapp.com/share/BQRA7GFY2EN#/screens).

**Comment by mapk** [comment]
I can't wait to see continued progress on this! It's looking good.
Since there's a "Needs Design Feedback" label here, I'll provide some. I'm fully aware this isn't really near completion, so take this with a grain of salt.

**1. The background grey**
I like this blueish-grey color, but maybe we should stick with the default grey background color provided in WordPress? `background-color: #f1f1f1`
Unless, of course, this grey is more consistent with Gutenberg greys.

**2. Block widths**
The width of the blocks within each Block Area appear a bit narrow.

Screen Shot 2019-04-05 at 2 17 26 PM


**3. Some blocks should be excluded**
I noticed I can add a "Page Break" block which doesn't seem to make sense in a siderbar or footer.

**4. Block toolbar**
There should be more padding above the first block in the area. Right now the block toolbar extends into the accordion title which doesn't seem right.

Screen Shot 2019-04-05 at 2 19 32 PM


**5. Block appender**
Just noting the block appender should be included in each area.

Let's remove the design feedback label for now since this is also being further discussed in the RFC. Once there's more to review, we can add the label back.

**Comment by youknowriad** [comment]
Thanks for the great feedback @mapk 

1. Updated

2. That's because the default block editor contain padding to show the block movers and there's probably some extra unnecessary padding there. 

3. Definitely, I think there needs to be APIs/rules to show/exclude blocks from post content editor and from widget editors. I think this deserves a big discussion to define it properly. For example, maybe all blocks depending on the `wp-editor` script could be considered post-specific? 

4. I noticed that the toolbar don't show up consistently. Not sure why yet but it deserves a follow-up

I also noticed that the block movers are not showing-up (not sure why yet).

I'm going to merge this PR. there's a lot of work remaining here, but this should give us the basis from which we will be able to iterate all together.

**Pull request merged**
